server/src/test/java/org/apache/druid/server/coordinator/simulate/CoordinatorSimulationBuilder.java

```diff
@@ -0,0 +1,549 @@
+      this.loadImmediately = loadImmediately;
+      this.autoSyncInventory = autoSyncInventory;
+
+      this.coordinatorConfig = new TestDruidCoordinatorConfig.Builder()
+          .withCoordinatorStartDelay(new Duration(1L))
+          .withCoordinatorPeriod(new Duration(DEFAULT_COORDINATOR_PERIOD))
+          .withCoordinatorKillPeriod(new Duration(DEFAULT_COORDINATOR_PERIOD))
+          .withLoadQueuePeonRepeatDelay(new Duration("PT0S"))
+          .withLoadQueuePeonType("http")
+          .withCoordinatorKillIgnoreDurationToRetain(false)
+          .build();
+
+      this.executorFactory = new ExecutorFactory(loadImmediately);
+      this.coordinatorInventoryView = autoSyncInventory
+                                      ? clusterInventory
+                                      : new TestServerInventoryView();
+      HttpClient httpClient = new TestSegmentLoadingHttpClient(
+          OBJECT_MAPPER,
+          clusterInventory::getChangeHandlerForHost,
+          executorFactory.create(1, ExecutorFactory.HISTORICAL_LOADER)
+      );
+
+      this.loadQueueTaskMaster = new LoadQueueTaskMaster(
+          null,
+          OBJECT_MAPPER,
+          executorFactory.create(1, ExecutorFactory.LOAD_QUEUE_EXECUTOR),
+          executorFactory.create(1, ExecutorFactory.LOAD_CALLBACK_EXECUTOR),
+          coordinatorConfig,
+          httpClient,
+          null
+      );
+
+      this.jacksonConfigManager = mockConfigManager();
+      setDynamicConfig(dynamicConfig);
+
+      this.lookupCoordinatorManager = EasyMock.createNiceMock(LookupCoordinatorManager.class);
+      mocks.add(jacksonConfigManager);
+      mocks.add(lookupCoordinatorManager);
+    }
+
+    private void setUp() throws Exception
+    {
+      EmittingLogger.registerEmitter(serviceEmitter);
+      historicalInventoryView.setUp();
+      coordinatorInventoryView.setUp();
+      lifecycle.start();
+      executorFactory.setUp();
+      leaderSelector.becomeLeader();
+      EasyMock.replay(mocks.toArray());
+    }
+
+    private void tearDown()
+    {
+      EasyMock.verify(mocks.toArray());
+      executorFactory.tearDown();
+      lifecycle.stop();
+    }
+
+    private void setDynamicConfig(CoordinatorDynamicConfig dynamicConfig)
+    {
+      this.dynamicConfig.set(dynamicConfig);
+    }
+
+    private JacksonConfigManager mockConfigManager()
+    {
+      final JacksonConfigManager jacksonConfigManager
+          = EasyMock.createMock(JacksonConfigManager.class);
+      EasyMock.expect(
+          jacksonConfigManager.watch(
+              EasyMock.eq(CoordinatorDynamicConfig.CONFIG_KEY),
+              EasyMock.eq(CoordinatorDynamicConfig.class),
+              EasyMock.anyObject()
+          )
+      ).andReturn(dynamicConfig).anyTimes();
+
+      EasyMock.expect(
+          jacksonConfigManager.watch(
+              EasyMock.eq(CoordinatorCompactionConfig.CONFIG_KEY),
+              EasyMock.eq(CoordinatorCompactionConfig.class),
+              EasyMock.anyObject()
```

Comment:
Suggestion: rather than mocking a config, add a constructor or builder to the config. A class that cannot be constructed (as with our config classes) is very tedious to use in tests. Mocking isn't the answer since, if there are methods that compute values, those methods also must be mocked.

Another solution is to use Guice. Use the new startup config builder and friends to pass in the set of properties, then ask the injector to create config instances. Going that route is a bit overkill, but it avoids the need to add constructors: we just use the Json config process we already have.

Reply:
> Suggestion: rather than mocking a config, add a constructor or builder to the config. A class that cannot be constructed (as with our config classes) is very tedious to use in tests. Mocking isn't the answer since, if there are methods that compute values, those methods also must be mocked.

I completely agree, most of our configs have private fields, no setters and no builders.